WalletWasabi/WabiSabi/Models/Decomposition/BaseDenomGenerator.cs

```diff
@@ -0,0 +1,40 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Models.Decomposition
+{
+	public static class BaseDenominationGenerator
+	{
+		private const long MaxSats = 2099999997690000;
+
+		private static IEnumerable<long> Multiple(IEnumerable<int> coefficients, IEnumerable<long> values)
+		{
+			foreach (var v in values)
+			{
+				foreach (var c in coefficients)
+				{
+					var x = c * v;
+					if (x > MaxSats)
+					{
+						break;
+					}
+					yield return x;
+				}
+			}
+		}
```

Comment:
IIRC LINQ is preferred in Wasabi, I think both are fine (untested):

```suggestion
		private static IEnumerable<long> Multiple(IEnumerable<int> coefficients, IEnumerable<long> values)
		    => values.SelectMany(v => coefficients.Select(c => c * v)).Where(x => x < MaxSats))
```

It would also make sense to return `IEnumerable<Money>` already here and just eliminate the multiples helper, C# is more powerful at declarative things like that.

Reply:
ACK. Not x <= MaxSats?

Reply:
> It would also make sense to return IEnumerable<Money> already here and just eliminate the multiples helper, C# is more powerful at declarative things like that.

First part ACK, second part I was not sure what you meant, I did that.